jwst/outlier_detection/utils.py

```diff
@@ -20,177 +26,362 @@
-            example_model = resampled_models.borrow(0)
-            shp = example_model.data.shape
-            dtype = example_model.data.dtype
-            nsections, section_nrows = _compute_buffer_indices(example_model, buffer_size)
-            resampled_models.shelve(example_model, modify=False)
-            del example_model
+        median_data = median_computer.compute_median()
+        median_computer.cleanup()
 
-        # get spatial sections of library and compute timewise median, one by one
-        resampled_sections = _get_sections(resampled_models, nsections, section_nrows, shp[0])
-        median_image_empty = np.empty(shp, dtype) * np.nan
-        return _create_median(resampled_sections, resampled_models, weight_thresholds, median_image_empty)
+    if save_intermediate_results:
+        # Save median model to fits
+        median_model = datamodels.ImageModel(median_data)
+        median_model.update(drizzled_model)
+        median_model.meta.wcs = median_wcs
+        save_median(median_model, make_output_path, resamp.asn_id)
+        del median_model
+    del drizzled_model
 
+    return median_data, median_wcs
 
-def _get_sections(library, nsections, section_nrows, imrows):
-    """Iterator to return sections from a ModelLibrary.
 
-    Parameters
-    ----------
-    library : ModelLibrary
-        The input data models.
+class DiskAppendableArray:
+    """
+    Creates a temporary file to which to append data, in order to perform
+    timewise operations on a stack of input images without holding all of them
+    in memory.
+    
+    This class is purpose-built for the median computation during outlier detection 
+    and is not very flexible. It is assumed that each data array passed to `append`
+    represents the same spatial segment of the full dataset. It is also assumed that 
+    each data array passed to `append` represents only a single instant in time; 
+    the append operation will stack them along a new axis.
+    
+    The `read` operation is only capable of loading the full array back into memory.
+    When working with large datasets that do not fit in memory, the
+    required workflow is to create many DiskAppendableArray objects, each holding
+    a small spatial segment of the full dataset.
+    """
+    
+    def __init__(self, slice_shape, dtype, tempdir):
+        """
+        Parameters
+        ----------
+        slice_shape : tuple
+            The shape of the spatial section of input data to be appended to the array.
+
+        dtype : str
+            The data type of the array. Must be a valid numpy array datatype.
+
+        tempdir : str
+            The directory in which to create the temporary files.
+            Default is the current working directory.
+        """
+        if len(slice_shape) != 2:
+            raise ValueError(f"Invalid slice_shape {slice_shape}. Only 2-D arrays are supported.")
+        self._temp_file, self._filename = tempfile.mkstemp(dir=tempdir)
```

Comment:
`mkstemp` returns an open file and a filename. I don't see where `_temp_file` is closed so it looks like this is still keeping 1 file open for each `DiskAppendableArray` (and relying on `__del__` to close the file).

To avoid this (and I think simplify things) how about requiring `DiskAppendableArray` accept a filename? `OnDiskMedian` can then be updated to create filenames for each array using `self._temp_path / f"{index}.bin"` (with the index of the `DiskAppendableArray` instance in the `temp_arrays` list). These files could then all be deleted after `temp_arrays` is no longer needed (in `OnDiskMedian.cleanup`) or allow the files to be deleted when the `TemporaryDirectory.cleanup` is called (I don't think this produces any warnings or errors and just silently cleans up all the files).

Reply:
Good suggestion.  I put this into the latest commit - let me know if it resolves your comment